sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs

```diff
@@ -0,0 +1,126 @@
+            KnownType.System_NotImplementedException,
+            KnownType.System_NotSupportedException);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var methods = classDeclaration.Members.OfType<MethodDeclarationSyntax>().ToList();
+
+                    var alreadyHandledMethods = new HashSet<MethodDeclarationSyntax>();
+
+                    foreach (var method in methods)
+                    {
+                        if (alreadyHandledMethods.Contains(method))
+                        {
+                            continue;
+                        }
+
+                        var duplicates = methods.Where(m => AreDuplicates(method, m, c.SemanticModel)).ToList();
+                        if (duplicates.Count > 0)
+                        {
+                            alreadyHandledMethods.Add(method);
+                            alreadyHandledMethods.UnionWith(duplicates);
+
+                            foreach (var duplicate in duplicates)
+                            {
+                                c.ReportDiagnostic(Diagnostic.Create(rule, duplicate.Identifier.GetLocation(),
+                                    additionalLocations: new[] { method.Identifier.GetLocation() },
+                                    messageArgs: method.Identifier.ValueText));
+                            }
+                        }
+                    }
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private bool AreDuplicates(MethodDeclarationSyntax firstMethod, MethodDeclarationSyntax secondMethod,
+            SemanticModel model)
+        {
+            if (firstMethod == secondMethod)
+            {
+                return false;
+            }
+
+            if (firstMethod.Body != null &&
+                secondMethod.Body != null)
+            {
+                return firstMethod.Body.Statements.Count >= 2 &&
+                    firstMethod.Identifier.ValueText != secondMethod.Identifier.ValueText &&
+                    firstMethod.Body.IsEquivalentTo(secondMethod.Body, false);
+            }
+
+            return false;
+        }
+
+        private bool IsIgnoredThrowStatement(StatementSyntax statement, SemanticModel model)
+        {
+            var throwStatement = statement as ThrowStatementSyntax;
+            if (throwStatement == null ||
+                !throwStatement.Expression.IsKind(SyntaxKind.ObjectCreationExpression))
+            {
+                return false;
+            }
+
+            var symbol = model.GetSymbolInfo(throwStatement.Expression).Symbol?.ContainingType;
+            return symbol.IsAny(ThrowStatementTypesToIgnore);
+        }
+
+        private class MethodDuplicates
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove the unused private type 'MethodDuplicates'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)